overlord/snapstate/snapstate.go

```diff
@@ -1978,10 +1992,135 @@ func updateManyFiltered(ctx context.Context, st *state.State, names []string, re
+	// taskset with essential snaps (snapd, kernel, gadget and the model base)
+	essentialUpdated, essentialTss, err := updateFunc(essential)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// taskset with non-essential snaps (apps and their bases)
+	nonEssentialUpdated, nonEssentialTss, err := updateFunc(nonEssential)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	allUpdated := append(essentialUpdated, nonEssentialUpdated...)
+
+	// if snapd is in the essential snaps set, the non-essentials must wait for it
+	if strutil.ListContains(allUpdated, "snapd") {
+		snapdTss, err := findTasksetForSnap(essentialTss.Refresh, "snapd")
+		if err != nil {
+			return nil, nil, err
+		}
+
+		// make non-essential snaps also wait for snapd
+		snapdEndTask := snapdTss.MaybeEdge(EndEdge)
+		if snapdEndTask == nil {
+			return nil, nil, fmt.Errorf("internal error: cannot find last task in snapd's update taskset")
+		}
+
+		for _, ts := range nonEssentialTss.Refresh {
+			startTask := ts.MaybeEdge(BeginEdge)
+			if startTask == nil {
+				return nil, nil, fmt.Errorf("internal error: cannot find first task in snap's taskset")
+			}
+			startTask.WaitFor(snapdEndTask)
+		}
+	}
+
+	// add dependencies between apps and the boot base, if required
+	for _, base := range essential {
+		if base.Type() != snap.TypeBase {
+			continue
+		}
+
+		for _, app := range nonEssential {
+			if app.SnapBase() != base.InstanceName() {
+				continue
+			}
+
+			baseTaskset, err := findTasksetForSnap(essentialTss.Refresh, base.InstanceName())
+			if err != nil {
+				return nil, nil, err
+			}
+
+			appTaskset, err := findTasksetForSnap(nonEssentialTss.Refresh, app.InstanceName())
+			if err != nil {
+				return nil, nil, err
+			}
+
+			if baseTaskset == nil || appTaskset == nil {
+				// one of the snaps is not being updated
+				continue
+			}
+
+			// make the app wait for the base
+			baseEndTask := baseTaskset.MaybeEdge(EndEdge)
+			if baseEndTask == nil {
+				return nil, nil, fmt.Errorf("internal error: cannot find last task in base's update taskset")
+			}
+			appStartTask := appTaskset.MaybeEdge(BeginEdge)
+			if appStartTask == nil {
+				return nil, nil, fmt.Errorf("internal error: cannot find first task in snap's taskset")
+			}
+
+			appStartTask.WaitFor(baseEndTask)
+		}
+	}
+
+	return allUpdated, &UpdateTaskSets{
+		// only non-essential snaps can trigger pre-downloads
+		PreDownload: nonEssentialTss.PreDownload,
+		Refresh:     append(essentialTss.Refresh, nonEssentialTss.Refresh...),
+	}, nil
+}
+
+func findTasksetForSnap(tss []*state.TaskSet, name string) (*state.TaskSet, error) {
+	for _, ts := range tss {
+		for _, t := range ts.Tasks() {
+			var snapsup SnapSetup
+			err := t.Get("snap-setup", &snapsup)
+			if err != nil {
+				if errors.Is(err, state.ErrNoState) {
+					continue
+				}
+				return nil, err
+			}
+			if snapsup.InstanceName() != name {
+				break
+			}
+			return ts, nil
+		}
+	}
+
+	return nil, nil
```

Comment:
Maybe name this function to something that implies we could return `nil`, even in the absence of an error?

Reply:
Done, thanks!